src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -0,0 +1,1146 @@
+        /// may not be in the actual workspace yet.
+        /// </summary>
+        private readonly Dictionary<DocumentId, FileChangeWatcher.IFileWatchingToken> _fileWatchingTokens = new Dictionary<DocumentId, FileChangeWatcher.IFileWatchingToken>();
+
+        /// <summary>
+        /// A file change context used to watch source files and additional files for this project. It's automatically set to watch the user's project
+        /// directory so we avoid file-by-file watching.
+        /// </summary>
+        private readonly FileChangeWatcher.IContext _documentFileChangeContext;
+
+        /// <summary>
+        /// A file change context used to watch metadata and analyzer references.
+        /// </summary>
+        private readonly FileChangeWatcher.IContext _fileReferenceChangeContext;
+
+        private readonly DocumentCollection _sourceFiles;
+        private readonly DocumentCollection _additionalFiles;
+
+        internal VisualStudioProject(VisualStudioWorkspaceImpl workspace, HostDiagnosticUpdateSource hostDiagnosticUpdateSource, ProjectId id, string projectUniqueName, string language, string directoryNameOpt)
+        {
+            _workspace = workspace;
+            _hostDiagnosticUpdateSource = hostDiagnosticUpdateSource;
+
+            Id = id;
+            Language = language;
+            _displayName = projectUniqueName;
+            _projectUniqueName = projectUniqueName;
+
+            if (directoryNameOpt != null)
+            {
+                _documentFileChangeContext = _workspace.FileChangeWatcher.CreateContextForDirectory(directoryNameOpt);
+            }
+            else
+            {
+                _documentFileChangeContext = workspace.FileChangeWatcher.CreateContext();
+            }
+
+            _documentFileChangeContext.FileChanged += FileChangeContext_FileChanged;
+
+            // TODO: set this to watch the NuGet directory or the reference assemblies directory; since those change rarely and most references
+            // will come from them, we can avoid creating a bunch of explicit file watchers.
+            _fileReferenceChangeContext = workspace.FileChangeWatcher.CreateContext();
+
+            _sourceFiles = new DocumentCollection(this, (s, d) => s.ContainsDocument(d), (w, d) => w.OnDocumentAdded(d), (w, documentId) => w.OnDocumentRemoved(documentId));
+            _additionalFiles = new DocumentCollection(this, (s, d) => s.ContainsAdditionalDocument(d), (w, d) => w.OnAdditionalDocumentAdded(d), (w, documentId) => w.OnAdditionalDocumentRemoved(documentId));
+        }
+
+        public ProjectId Id { get; }
+
+        private void ChangeProjectProperty<T>(ref T field, T newValue, Func<Solution, Solution> withNewValue, Action<Workspace> changeValue)
+        {
+            lock (_gate)
+            {
+                // If nothing is changing, we can skip entirely
+                if (object.Equals(field, newValue))
+                {
+                    return;
+                }
```

Comment:
I think this might be incorrect when combined with batching. Suppose the current value of the field is `A` and we batched a change to `B` and then another change to `A`. When the batch is applied the final value should be `A`, but this check causes the second batched update to be ignored, so the final value will be `B`.